ouroboros-network/src/Ouroboros/Network/Mux.hs

```diff
@@ -57,11 +60,21 @@ muxStart udesc bearer rescb_m = do
                ]
     mjobs <- sequence [ mpsJob cnt pmss ptcl
                       | ptcl <- [minBound..maxBound] ]
-    aids <- mapM async $ jobs ++ concat mjobs
+    aids <- runConcurrentlyA (jobs ++ concat mjobs)
     muxBearerSetState bearer Mature
     watcher aids
 
   where
+    -- like @mapConcurrently@, but it will return the @Async@s, instead of
+    -- waiting for the results.
+    runConcurrentlyA :: [m x] -> m [Async m x]
+    runConcurrentlyA = go []
+      where
+        go !acc []             = return acc
+        go !acc (mact : macts) = do
+          aid <- mask $ \unmask -> async (unmask mact) `catch` (\(e :: SomeAsyncException) -> forM_ acc cancel >> throwM e)
```

Comment:
I'd suggest not doing the cancel in the exception handler here. Instead, `traverse async :: [m a] -> m [Async m a]` while masked, and then take care of cancelling them on an exception here, rather than in each one. Something like...

```
mask $ \restore -> do
  asyncJobs <- traverse async jobs
  restore (k asyncJobs) `onException` forM_ asyncJobs cancel
  forM_ asyncJobs cancel
```

Reply:
That's true; it's better to attach error handlers when things are masked, and use `waitAnyCatch` in the `watcher`.

Reply:
I think we also need to `restore` the exceptions in the jobs, e.g.
```
mask $ \restore -> do
  asyncJobs <- traverse (async . restore) jobs
  restore (k asyncJobs) `onException` forM_ asyncJobs cancel
  forM_ asyncJobs cancel
```